lonestar/analytics/cpu/triangles/Triangles.cpp

```diff
@@ -354,46 +354,98 @@ void edgeIteratingAlgo(Graph& graph) {
   std::cout << "NumTriangles: " << numTriangles.reduce() << "\n";
 }
 
-void makeGraph(Graph& graph, const std::string& triangleFilename) {
-  typedef galois::graphs::FileGraph G;
-  typedef G::GraphNode N;
-
-  G initial, permuted;
-
+//! Sorts read graph by degree (high degree nodes are reindexed to beginning)
+void makeSortedGraph(Graph& graph) {
+  // read original graph
+  galois::graphs::FileGraph initial;
   initial.fromFileInterleaved<void>(inputFilename);
 
-  // Getting around lack of resize for deque
-  std::deque<N> nodes;
-  std::copy(initial.begin(), initial.end(), std::back_inserter(nodes));
-
-  /* Sort by degree:
-   *  DegreeLess: Sorts in the ascending order of node degrees
-   *  DegreeGreater: Sorts in the descending order of the node degrees
-   *
-   *  The order of sorting has a huge impact on performance
-   *  For this algorithm, sorting in descending order delivers the
-   *  best performance due to the way ties are broken.
-   */
-  galois::ParallelSTL::sort(nodes.begin(), nodes.end(),
-                            DegreeGreater<G>(initial));
-
-  std::deque<N> p;
-  std::copy(nodes.begin(), nodes.end(), std::back_inserter(p));
-  // Transpose
-  size_t idx = 0;
-  for (N n : nodes) {
-    p[n] = idx++;
+  size_t numGraphNodes = initial.size();
+  // create node -> degree pairs
+  using DegreeNodePair = std::pair<uint64_t, uint32_t>;
+  std::vector<DegreeNodePair> dnPairs(numGraphNodes);
+  galois::do_all(
+    galois::iterate((size_t)0, numGraphNodes),
+    [&] (size_t nodeID) {
+      size_t nodeDegree = std::distance(initial.edge_begin(nodeID),
+                                        initial.edge_end(nodeID));
+      dnPairs[nodeID] = DegreeNodePair(nodeDegree, nodeID);
+    },
+    galois::loopname("CreateDegreeNodeVector")
+  );
+
+  // sort by degree (first item)
+  std::sort(dnPairs.begin(), dnPairs.end(), std::greater<DegreeNodePair>());
+
+  // create mapping, get degrees out to another vector to get prefix sum
+  std::vector<uint32_t> oldToNewMapping(numGraphNodes);
+  std::vector<uint64_t> newPrefixSum(numGraphNodes);
+  galois::do_all(galois::iterate((size_t)0, numGraphNodes),
+    [&] (size_t index) {
+      // save degree, which is pair.first
+      newPrefixSum[index] = dnPairs[index].first;
+      // save mapping; original index is in .second, map it to current index
+      oldToNewMapping[dnPairs[index].second] = index;
+    },
+    galois::loopname("CreateRemappingGetPrefixSum")
+  );
+
+  // get prefix sum
+  for (size_t i = 1; i < numGraphNodes; i++) {
```

Comment:
This should be parallel. We might have a parallel prefix sum routine in our library. Check with @chenxuhao. Otherwise, use on_each and parallelize it (you can merge it with the previous do_all as well).

Reply:
pinged Xuhao elsewhere, will see what he says.

Reply:
Will be moving the parallel prefix sum to ParallelSTL as a function.